packages/datasource-customizer/src/decorators/chart/collection.ts

```
@@ -0,0 +1,36 @@
+import { Caller, Chart, CollectionSchema, CompositeId } from '@forestadmin/datasource-toolkit';
+import { CollectionChartDefinition } from './types';
+import CollectionChartContext from './context';
+import CollectionDecorator from '../collection-decorator';
+import ResultBuilder from './result-builder';
+
+export default class ChartCollectionDecorator extends CollectionDecorator {
+  private charts: Record<string, CollectionChartDefinition> = {};
+
+  addChart(name: string, definition: CollectionChartDefinition) {
+    if (this.schema.charts.includes(name)) {
+      throw new Error(`Chart '${name}' already exists.`);
+    }
+
+    this.charts[name] = definition;
+    this.markSchemaAsDirty();
+  }
+
+  override async renderChart(caller: Caller, name: string, recordId: CompositeId): Promise<Chart> {
+    if (this.charts[name]) {
+      const context = new CollectionChartContext(this, caller, recordId);
+      const resultBuilder = new ResultBuilder();
+
+      return this.charts[name](context, resultBuilder);
```

If the definition return nothing, the chart does not work and we do not have error. Can we throw an error to help the user ?

---

If that's OK with you, I would love to do a PR which handle validating the output of all customer code in one go.

We have the same issue with... everything! (action, datasource charts, fields, segments, ...)